packages/flutter_tools/lib/src/debug_adapters/flutter_adapter.dart

```diff
@@ -83,11 +83,41 @@ class FlutterDebugAdapter extends DartDebugAdapter<FlutterLaunchRequestArguments
   @override
   bool get terminateOnVmServiceClose => false;
 
+  /// Whether or not the user requested to debug (via noDebug).
```

Comment:
Would it also be correct for this to be a getter, like:

```
bool get debug {
  final DartCommonLaunchAttachRequestArguments args = this.args;
  if (args is! FlutterAttachRequestArguments) {
    return false;
  }
  return !(args.noDebug ?? false);
}
```

It took me quite a while to understand what this code was doing (and maybe I still misunderstood).

Reply:
Almost, yeah. It would actually be more like:

```dart
bool get debug {
    final args = this.args;
    final noDebug = args is FlutterLaunchRequestArguments
        ? args.noDebug
        : args is FlutterAttachRequestArguments
            ? args.noDebug
            : false; // Should not happen, args is always FlutterLaunchRequestArguments or FlutterAttachRequestArguments
    return !(noDebug ?? false);
  }
```

It's a bit odd because `noDebug` is not on the base arguments class because for Dart it's not valid to do noDebug for Attach (because we need a VM Service connection to attach). Flutter actually allows us to attach without a VM Service though, so `noDebug` ends up on its attach too.

A nicer way might be to add a common case for Flutters args (like `DartCommonLaunchAttachRequestArguments`). However, looking again at [the DAP spec](https://microsoft.github.io/debug-adapter-protocol/specification) it also does not list `noDebug` as a valid option for attach anyway, so now I'm unsure whether VS Code would actually honour the difference between Run/Debug when using a launch config with `"request": "attach"`.

Let me do some more testing of this.. if VS Code doesn't support it, then we should remove it from here too which simplifies things a little more (to something like your example, but `return true` is `args is! FlutterLaunchRequestArguments`).

> It took me quite a while to understand what this code was doing (and maybe I still misunderstood).

Anything in particular? I should make the cod and/or comments clearer if this is the case :-)

Reply:
> Anything in particular? I should make the code and/or comments clearer then :-)

Mostly the late initialization of the `debug` property and the fact that the arg is `noDebug` but the property is `debug`. I think if it's refactored to be a getter it will improve the maintainability.

Reply:
> and the fact that the arg is `noDebug` but the property is debug

Yeah, this is pretty confusing. `noDebug` comes from DAP so we can't change it.. it's a negative and it's optional, defaulting to false (meaning debug=true). I think wrapping it up in a getter so this inversion is only in a single place makes a lot of sense. I'll do this tomorrow, as well as remove it from `attach` if it turns out that VS Code doesn't pass it (if it does pass it, I'll file a DAP issue asking for clarification on its support).

Reply:
I've removed `noDebug` from attach now, and added a getter for `debug` that hopefully makes things clearer. Let me know if not!